Added beforematch explainer (#117)
WICG/display-locking: explainer-beforematch.md

# The `beforematch` event

### Summary (TL;DR)

> The `beforematch` event is fired at render timing in the following situations:

Reviewer: ```suggestion
The `beforematch` event allows developers to display content to the user in response to to find-in-page or fragment navigation actions. It is fired just before these actions, and at render timing. The applicable actions are:
```

* There is a new find-in-page active match, in which case the event is fired

>   with the matched element as the target.

Author: I think it should be the element that contains the text node

* There is a fragment-link navigation, in which case the event is fired with

>   the matched fragment as the target.

Author: I added a blurb about this.

* There is a scroll-to-text navigation, in which case the event is fired with
  the matched element as the target.

> 

Reviewer: The use case for this event is to allow developers to help users find hidden content on the page. Hidden content includes collapsed sections, `visibility: hidden` subtrees, and `subtree-visibility: hidden-matchable` [link to spec] subtrees.

Author: We don't yet have a spec for hidden-matchable, but I'll add that.

### Motivation

With the evolution of the web, there are always new and interesting ways that
developers choose to organize the information on their pages. Some of these
approaches (e.g. the common case of text scrolling), yield themselves naturally to
user-agent features like find-in-page. This is not an accident, since
find-in-page was designed with common use-cases in mind.

However, there are other interesting approaches, such as collapsed sections of
text, do not work well with user-agent features since the page does not get any
indication that the user initiated a find-in-page request, or fragment or

> scroll-to-text navigation.

Author: Yeah that's a consequence of rewriting the same sentence several times :) Thanks for the better suggestion.

The `beforematch` event is a signal that is a step in the direction that allows
us to leverage information that the user-agent has to assist the page in making

> the user experience great. Specifically, with future additions to

Reviewer: Suggest deleting the second sentence

Reviewer: Because it de-emphasizes the independent nature of this feature. It has value for collapsed sections already, irrespective of display locking. In fact, the parters who are testing out this use case in our OT don't anticipate any benefit to their use case of the performance aspect of display locking, just the part about discoverability in UA algorithms.

Reviewer 2: Just chime in here. In the Motivation section of the readme, it is mentioned that

> For these reasons, web developers need ways to reduce loading and rendering time of web apps that have a lot of non-visible DOM.

However @chrishtr mentioned above that

> In fact, the parters who are testing out this use case in our OT don't anticipate any benefit to their use case of the performance aspect of display locking

It seems a bit contradictory to me. Are we implying that renderSubtree is focusing more on solving the discoverability problem? 

P.S I am an OT user and yes I didn't observe much performance improvement after applying the display locking to our long list

Author: I think @chrishtr comment is about the beforematch event specifically. Display locking addresses a couple of use-cases, one being performance and another being discoverability. When it comes to the beforematch event, the important part of display locking is discoverability.

I'm disappointed to hear that you didn't observe performance benefits from display locking. If you don't mind and have the time to do it, can you file an issue in this repo with some details of how you used it, what you expected to see, and what happened instead? This is an important use-case for us, and we'd like to understand how we can improve performance.

It's also possible that there may be something we overlooked in the implementation which prevents you from seeing the performance results. In either case, the more information you can give us, the better

display-locking functionality, it will be possible to process text for find-in-page
match in sections that are not visible. In turn, the `beforematch` event will be

> fired on hidden (a.k.a. collapsed) sections, allowing the developer to unhide

Reviewer 3: Maybe it would be good to define what collapsed sections are? Are there any "collapsed" things other than display locking that this would be useful for?

Author: I added a footnote about hidden but matchable content. I don't really know of any real-world examples other than subtree-visibility though. Maybe Chris knows.

the section. The net effect is that the user is able to use find-in-page to find
content in a collapsed section -- something that is not currently possible.

Even without display-locking features, `beforematch` is useful signal to the
page which allows custom styling of the matched element, which is now only
possible with approximations from scroll positions and intersection
observations.

### Use Cases

> #### Example 1

Reviewer: Make this the last example. Add a first example that is about fragment navigation, both with anchor and text fragment.

Author: I think the fragment navigation and the text fragment thing really use the same example.. since it expands the section in the same way that find in page would, so I just added a couple of more paragraphs to that to explain (and reordered the examples)

}
.highlighted {
  background: cornsilk;
  transition: background 0s;
}
</style>

<div class=section>Lorem ipsum dolor sit amet</div>
<div class=section>consectetur adipiscing elit</div>
<div class=section>Sed augue lacus</div>

<script>
function highlight(e) {
  e.target.classList.add("highlighted");
  requestAnimationFrame(
    () => requestAnimationFrame(
      () => e.target.classList.remove("highlighted")));
}

document.querySelectorAll(".section").forEach(item => {
  item.addEventListener("beforematch", highlight);
});
</script>
```

In this example, a highlighted entry gains a corksilk background color, which
fades over approximately two seconds. It's a subtle effect that grabs the user's
attention but is not too intrusive. 

We use a `beforematch` event to highlight the section. Note that the highlighted
element is the target element of the `beforematch` event, but we can also modify
any related style or DOM based on the target's location.

The effect of this example is that the section is briefly highlighted in
cornsilk on an active find-in-page match. This effect happens in addition to the
user-agent highlight the found match.

#### Example 2
```html
<style>
.collapsed {

>   subtree-visibility: hidden-matchable;

Reviewer 3: Are there other subtree-visibility values that are functionally the same but wouldn't get the beforematch event fired on them?

Reviewer: Yes - subtree-visibility: hidden.

}
</style>

Please explore the following sections:
<h1>Introduction</h1> <div class=collapsed>...</div>
<h1>Thesis</h1>       <div class=collapsed>...</div>
<h1>Conclusion</h1>   <div class=collapsed>...</div>

<script>
function expand(e) {
  e.target.classList.remove("collapsed");
}

document.querySelectorAll(".collapsed").forEach(item => {
  item.addEventListener("beforematch", expand);
});
</script>
```

In this example, most of the content of the page is hidden in collapsed section.

> It uses a display-locking property called `subtree-visibility` to hide the

Reviewer: ```suggestion
It uses the upcoming `subtree-visibility` CSS property to hide the
```

Reviewer: Also link to the privacy section of the spec for text fragment navigation for more considerations that may be relevant.

Author: turns out there was not much there

content while letting it be searchable by find-in-page. When a match is found,
and `beforematch` event is fired, we expand the section by removing the
collapsed class.

Note that the net effect of this is that the user is able to find matches in
collapsed sections which are automatically expanded. This is possible due to
both display-locking functionality and the `beforematch` event.

Also note that the adoption of the `beforematch` in these use-cases should be
straight-forward, since the typical page that provides content in collapsed
sections already has an event handler to expand and show the section. With the
`beforematch` event, we can reuse the same handler to expand the section.

### Privacy Concerns

> Note that this event exposes more information to the origin that would otherwise

Reviewer 3: Should "origin" be "page" instead to make the use of vocabulary more consistent? Or is there a distinction?

> be available. In particular, the page can known which section of text was found

Reviewer 3: known -> know

> using find-in-page, fragment navigation, and scroll-to-text navigation.

Reviewer: Add: the developer may also be able to find out some information about what the user typed into a find-in-page UX based on which elements receive an event. Likewise, for scroll to text fragment, the developer may be able to learn more about the content of the fragment search terms based on which section of the page receives an event.

# The `beforematch` event

### Summary (TL;DR)

The `beforematch` event allows developers to display content to the user in
response to to find-in-page or fragment navigation actions. It is fired just
before these actions, and at render timing. The applicable actions are the
following:

* There is a new find-in-page active match, in which case the event is fired
  with the matched element as the target.
* There is a fragment-link navigation, in which case the event is fired with
  the matched fragment as the target.
* There is a scroll-to-text navigation, in which case the event is fired with
  the matched element as the target.

> Note that the 'matched element' in this document refer to one of the following:

Reviewer: ```suggestion
Note that the 'matched element' in this document refers to one of the following:
```

* the element that contains the text which was matched by find-in-page or
  scroll-to-text navigation.
* the element which was selected for the fragment link navigation.

The use case for this event is to allow developers to help users find hidden
content on the page. Hidden content includes collapsed sections via clipping,

> `visibility: hidden` subtree, and [`subtree-visibility:

Reviewer: ```suggestion
`visibility: hidden` subtrees, and [`subtree-visibility:
```

hidden-matchable`](https://github.com/WICG/display-locking) subtrees.

### Motivation

With the evolution of the web, there are always new and interesting ways that
developers choose to organize the information on their pages. Some of these

> approaches (e.g. the common case of text scrolling), yield themselves naturally to

Reviewer: ```suggestion
approaches (e.g. the common case of text scrolling), lend themselves naturally to
```

user-agent features like find-in-page. This is not an accident, since
find-in-page was designed with common use-cases in mind.

However, other approaches like collapsed sections of text do not work well with
user-agent features since the page does not get any indication that the user
initiated a find-in-page request, fragment navigation, or scroll-to-text
navigation.

The `beforematch` event is a step in the direction that allows developers to
leverage information that the user-agent already has to make these search and
navigation experiences great. Specifically, with [hidden but
matchable](#footnotes) content, it will be possible to process text for
find-in-page match in sections that are not visible. In turn, the `beforematch`
event will be fired on hidden (a.k.a. collapsed) sections, allowing the
developer to unhide the section. The net effect is that the user is able to use

> find-in-page to find content in a collapsed section -- something that is not

Reviewer: ```suggestion
find-in-page or link navigation to find content in collapsed sections -- something that is not
```

something that is not currently possible.

> Even without hidden but matchable features, `beforematch` is useful signal to the

Reviewer: ```suggestion
Even without hidden but matchable features, `beforematch` is a useful signal to the
```

page which allows custom styling of the matched element, which is now only
possible with approximations from scroll positions and intersection
observations.

### Use Cases

#### Example 1
```html
<style>
.collapsed {
  subtree-visibility: hidden-matchable;
}
</style>

Please explore the following sections:
<h1>Introduction</h1> <div class=collapsed>lorem ipsum ...</div>
<h1>Thesis</h1>       <div class=collapsed>dolor sit amet ...<div>
<h1>Conclusion</h1>   <div class=collapsed>consectetur adipiscing ...</div>

<script>
function expand(e) {
  e.target.classList.remove("collapsed");
}

document.querySelectorAll(".collapsed").forEach(item => {
  item.addEventListener("beforematch", expand);
});
</script>
```

> In this example, most of the content of the page is hidden in collapsed section.

Reviewer: ```suggestion
In this example, most of the content of the page is hidden in collapsed sections.
```

It uses the upcoming `subtree-visibility` CSS property to hide the
content while letting it be searchable by find-in-page. When a match is found,
and `beforematch` event is fired, we expand the section by removing the
collapsed class.

Note that the net effect of this is that the user is able to find matches in
collapsed sections which are automatically expanded. This is possible due to
both hidden-but-matchable content and the `beforematch` event.

The same effect occurs when doing a fragment link navigation to any element
within the collapsed section. Likewise, scroll-to-text navigation would expand
relevant sections. For instance, navigating to
`example.com/page.html#:~:text=lorem` would expand the introduction section by
observing the `beforematch` event for that section as a result of the
scroll-to-text fragment navigation match.

> Also note that the adoption of the `beforematch` in these use-cases should be

Reviewer: ```suggestion
Also note that the adoption of the `beforematch` event in these use-cases should be
```

> Also note that the adoption of the `beforematch` event in these use-cases should be

Reviewer: ```suggestion
Also note that developer adoption of the `beforematch` event in these use-cases should be
```

sections already has an event handler to expand and show the section. With the
`beforematch` event, we can reuse the same handler to expand the section.

#### Example 2

```html
<style>
.section {
  transition: background 2s;
}
.highlighted {
  background: cornsilk;
  transition: background 0s;
}
</style>

<div class=section>Lorem ipsum dolor sit amet</div>
<div class=section>consectetur adipiscing elit</div>
<div class=section>Sed augue lacus</div>

<script>
function highlight(e) {
  e.target.classList.add("highlighted");
  requestAnimationFrame(
    () => requestAnimationFrame(
      () => e.target.classList.remove("highlighted")));
}

document.querySelectorAll(".section").forEach(item => {
  item.addEventListener("beforematch", highlight);
});
</script>
```

In this example, a highlighted entry gains a corksilk background color, which
fades over approximately two seconds. It's a subtle effect that grabs the user's
attention but is not too intrusive. 

We use a `beforematch` event to highlight the section. Note that the highlighted
element is the target element of the `beforematch` event, but we can also modify
any related style or DOM based on the target's location.

> The effect of this example is that the section is briefly highlighted in

Reviewer: This sentence is redundant.

> cornsilk on an active find-in-page match. This effect happens in addition to the

Reviewer: ```suggestion
cornsilk on an active find-in-page match. Any effect added by the developer happens in addition to the
```

user-agent highlight the found match.

### Privacy Concerns

Note that this event exposes more information to the page that would otherwise
be available. In particular, the page can know which section of text was found
using find-in-page, fragment navigation, and scroll-to-text navigation.

The developer may also be able to find out some information about what the user
typed into a find-in-page dialog based on which elements receive an event.
Likewise, for scroll-to-text navigation, the developer may be able to learn more
about the content of the fragment search terms based on which section of the
page receives an event.

We believe that the benefit of providing this information to the site outweighs
the potential risks. Moreover, the lower granularity information can already be
approximated from scroll position and intersection observerations on the page.

> The privacy aspect of the event should be discussed in more detail.

Reviewer: ```suggestion
The privacy aspect of the event should be discussed in more detail via formal security reviews.
```

### Footnotes

**Hidden but matchable** content refers to an idea that although some content
may be hidden from the user, it may still be useful to have that content be
searchable.  One such approach is [`subtree-visibility:

> hidden-matchable`](https://github.com/WICG/display-locking), but there may be

Reviewer: ```suggestion
hidden-matchable`](https://github.com/WICG/display-locking), but there are
```
